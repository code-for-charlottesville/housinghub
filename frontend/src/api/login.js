import axios from "axios";
import apiEndpoint from "./endpoint";

/**
 * makes POST request to /auth/login with username and password
 **/
export function postAuthLogin({ email, password }) {
  return axios
    .post(`${apiEndpoint}/auth/login`, {
      username: email,
      password,
    })
    .then((r) => {
      return Promise.resolve(r.data);
    })
    .catch((error) => {
      return Promise.resolve({
        error:
          (error.response.data && error.response.data.error) ||
          error.response.statusText,
      });
    });
}

/**
 * makes GET request to see if current token is valid
 **/
export function getStatus(jwt) {
  return axios
    .get(`${apiEndpoint}/auth/status`, {
      headers: { Authorization: `Bearer ${jwt}` },
    })
    .then((r) => {
      return Promise.resolve(r.data);
    })
    .catch((error) => {
      return Promise.resolve({
        error:
          (error.response.data && error.response.data.error) ||
          error.response.statusText,
      });
    });
}
