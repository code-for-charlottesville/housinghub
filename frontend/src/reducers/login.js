import axios from "axios";

const initialState = {
  error: "",
  isLoggedIn: false,
  fields: {
    email: "",
    password: "",
    rememberMe: true
  }
};

const appState = (state = initialState, action) => {
  switch (action.type) {
    case "LOGOUT":
      removeJwtFromLocalStorage();
      return Object.assign({}, state, {
        ...state,
        isLoggedIn: false,
        error: ""
      });
    case "SET_LOGIN_SUCCESS":
      if (state.fields.rememberMe) {
        setJwtInLocalStorage(action.jwt);
      }
      axios.defaults.headers.common["Authorization"] = `Beaerer ${action.jwt}`;
      return Object.assign({}, state, {
        ...state,
        isLoggedIn: true,
        error: ""
      });
    case "SET_LOGIN_ERROR":
      return Object.assign({}, state, {
        ...state,
        isLoggedIn: false,
        error: action.error
      });
    case "SET_LOGIN_FIELD":
      return Object.assign({}, state, {
        ...state,
        fields: {
          ...state.fields,
          [action.fieldName]: action.newValue
        }
      });
    default:
      return state;
  }
};

export default appState;

const LOCAL_STORAGE_JWT_KEY = "housinghub_token";
function setJwtInLocalStorage(jwt) {
  localStorage.setItem(LOCAL_STORAGE_JWT_KEY, jwt);
}

export function getJwtFromLocalStorage() {
  return localStorage.getItem(LOCAL_STORAGE_JWT_KEY);
}

export function removeJwtFromLocalStorage() {
  localStorage.removeItem(LOCAL_STORAGE_JWT_KEY);
  delete axios.defaults.headers.common["Authorization"];
}
