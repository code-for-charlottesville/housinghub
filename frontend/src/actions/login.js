import { postAuthLogin, getStatus } from "../api/login";
import { setLoading } from "./appState";
import store from "../reducers/index";
import {
  getJwtFromLocalStorage,
  removeJwtFromLocalStorage,
} from "../reducers/login";
import axios from "axios";
import { setUser } from "./user";

export function setLoginField(fieldName, newValue) {
  return {
    type: "SET_LOGIN_FIELD",
    fieldName,
    newValue,
  };
}

export function setLoginSuccess(jwt) {
  return {
    type: "SET_LOGIN_SUCCESS",
    jwt,
  };
}

export function setLoginFailure(error) {
  return {
    type: "SET_LOGIN_ERROR",
    error,
  };
}

export function logout() {
  return {
    type: "LOGOUT"
  };
}

/**
 * fires when user attempts to login
 **/
export function loginUser() {
  store.dispatch(setLoading(true));
  // Make API call to backend /auth/login
  postAuthLogin(store.getState().login.fields).then((response) => {
    store.dispatch(setLoading(false));
    if (response && response.jwt) {
      store.dispatch(setLoginSuccess(response.jwt));
      // set jwt status
      setJwtStatus(false);
    } else {
      store.dispatch(setLoginFailure(response.error));
    }
  });
}

// called when app starts up
export function setJwtStatus(checkLocalOnly = true) {
  // if there is a token in local storage, set as user logged in with that token
  if (getJwtFromLocalStorage() !== null || checkLocalOnly) {
    store.dispatch(setLoading(true));
    getStatus(getJwtFromLocalStorage()).then(r => {
      store.dispatch(setLoading(false));
      if (r.uid) {
        store.dispatch(setLoginSuccess(getJwtFromLocalStorage()));
        store.dispatch(setUser(r));
      } else {
        // else  token is invalid
        removeJwtFromLocalStorage();
      }
    });
  }
}
