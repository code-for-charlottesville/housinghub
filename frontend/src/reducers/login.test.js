import login from "./login";

describe("reducers/appState", () => {
  let testTable = [
    {
      name: "logout",
      action: {
        type: "LOGOUT",
      },
    },
    {
      name: "login success",
      action: {
        type: "SET_LOGIN_SUCCESS",
      },
    },
    {
      name: "login error",
      action: {
        type: "SET_LOGIN_ERROR",
        error: "sample error",
      },
    },
    {
      name: "set login field",
      action: {
        type: "SET_LOGIN_FIELD",
        fieldName: "testField",
        newValue: "new value",
      },
    },
  ];
  testTable.forEach((t) => {
    it(t.name, () => {
      expect(login(undefined, t.action)).toMatchSnapshot();
    });
  });
});
