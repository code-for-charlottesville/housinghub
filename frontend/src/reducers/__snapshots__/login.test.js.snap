// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/appState login error 1`] = `
Object {
  "error": "sample error",
  "fields": Object {
    "email": "",
    "password": "",
    "rememberMe": true,
  },
  "isLoggedIn": false,
}
`;

exports[`reducers/appState login success 1`] = `
Object {
  "error": "",
  "fields": Object {
    "email": "",
    "password": "",
    "rememberMe": true,
  },
  "isLoggedIn": true,
}
`;

exports[`reducers/appState logout 1`] = `
Object {
  "error": "",
  "fields": Object {
    "email": "",
    "password": "",
    "rememberMe": true,
  },
  "isLoggedIn": false,
}
`;

exports[`reducers/appState set login field 1`] = `
Object {
  "error": "",
  "fields": Object {
    "email": "",
    "password": "",
    "rememberMe": true,
    "testField": "new value",
  },
  "isLoggedIn": false,
}
`;
